docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,826 @@
+# Rework of Mesh*Routes matching
+
+* Status: accepted
+
+Technical story: https://github.com/kumahq/kuma/issues/10247
+
+## Context and Problem Statement
+
+At the moment when targeting `Mesh*Route` you need to put it in topLevel targetRef. 
+We want to iterate over policy design and move `Mesh*Route` to `spec.to[].targetRef` section follow model where `spec.targetRef`
+selects whole proxies, `spec.from[].targetRef` selects inbounds and `spec.to[].targetRef` selects outbounds
+
+## Decision Outcome
+
+- add possibility to put `Mesh*Routes` in `spec.to[].targetRef`
+
+## Considered options
+
+1. add possibility to put `Mesh*Routes` in `spec.to[].targetRef` -> `spec.targetRef` selects proxies, `spec.from[].targetRef` selects inbounds, `spec.to[].targetRef` selects outbounds.
+2. leave `Mesh*Routes` in top level targetRef as it is now, improve the semantics of `spec.targetRef` and introduce `spec.workloadSelector` to select whole proxies
+
+### Add possibility to put Mesh*Routes in `spec.to[].targetRef`
+
+#### Advantages
+
+- simplifying policy api as `spec.targetRef` will always pick proxy and `spec.to[].targetRef` section will select outgoing traffic
+- ability to set different confs on different proxy for the same route (overriding default conf for route without producer/consumer policies, on universal)
+- gives more flexibility as you can target multiple outbounds in single policy. Reducing number of consumer policies
+- more consistent with current approach
+- easier policy matching code
+- least problematic migration
+
+#### Disadvantages
+
+- can be harder to understand because `spec.targetRef` can be MeshSubset (which is optional and probably will rarely be used)
+- yet another migration for users (only for system policies as for namespaced policies this is already not allowed) 
+- mixing outbounds and routes in `spec.to[].targetRef`
+- hides the most often most important information under `to[]` adding noise when targeting Mesh in `spec.targetRef`
+- is it consistent in the case of MeshGateways? MeshHTTPRoutes aren't on outbound listeners in the case of MeshGateways, they're on inbound listeners
+
+### Leave Mesh\*Routes in top level targetRef as it is now and refine semantics of `spec.targetRef`
+
+Instead of having `targetRef` always refer to a workload,
+instead `targetRef` selects some part of the Envoy configuration of a workload,
+where that part is represented by a real Kuma resource.
+
+In fact `targetRef` is the wrong concept for selecting workloads,
+since "targetRef refers to a workload" means we aren't targeting
+an arbitrary `Kind` of resource but instead always a `Dataplane`.
```

Comment:
@slonka wrote:

> It feels like this https://github.com/kumahq/kuma/pull/10858/files#diff-95c559a0f6155d986d45f294106f1a5a789e99bd0c0c93dae4405724efc76f3cR47-R49 is the most compelling for me.

Reply:
Yea I liked this but in fact I kind of glossed over the gateway aspect of this. I had the `MeshHTTPRoute` case too far in the front of my mind and in general we would need a place to select whether it applies to gateways or mesh services, so it can't really be as simple as labels I guess.

Reply:
added